src/codecs/webp/vp8.rs

```diff
@@ -1382,6 +1424,113 @@ impl<R: Read> Vp8Decoder<R> {
+        let mut vws = [0u8; (8 + 1) * (8 + 1)];
+
+        let ylength = cmp::min(self.frame.chroma_height() as usize - mby*8, 8);
+        let xlength = cmp::min(self.frame.chroma_width() as usize - mbx*8, 8);
+
+        //left border
+        for y in 0usize..8 {
+            let (uy, vy) = if mbx == 0 || y >= ylength {
+                (127, 127)
+            } else {
+                let index = (mby * 8 + y) * w + ((mbx - 1) * 8 + 7);
+                (
+                    self.frame.ubuf[index],
+                    self.frame.vbuf[index]
+                )
+            };
+
+            uws[(y + 1) * stride] = uy;
+            vws[(y + 1) * stride] = vy;
+        }
+        //top border
+        for x in 0usize..8 {
+            let (ux, vx) = if mby == 0 || x >= xlength {
+                (129, 129)
+            } else {
+                let index = ((mby - 1) * 8 + 7) * w + (mbx * 8 + x);
+                (
+                    self.frame.ubuf[index],
+                    self.frame.vbuf[index]
+                )
+            };
+
+            uws[x + 1] = ux;
+            vws[x + 1] = vx;
+        }
+
+        //top left point
+        let (u1, v1) = if mby == 0 {
+            (127, 127)
+        } else if mbx == 0 {
+            (129, 129)
+        } else {
+            let index = ((mby - 1) * 8 + 7) * w + (mbx - 1) * 8 + 7;
+            if index >= self.frame.ubuf.len() {
+                (127, 127)
+            } else {
+                (
+                    self.frame.ubuf[index],
+                    self.frame.vbuf[index]
+                )
+            }
+        };
+
+        uws[0] = u1;
+        vws[0] = v1;
+
+        match mb.chroma_mode {
+            ChromaMode::DC => {
+                predict_dcpred(&mut uws, 8, stride, mby != 0, mbx != 0);
+                predict_dcpred(&mut vws, 8, stride, mby != 0, mbx != 0);
+            },
+            ChromaMode::V => {
+                predict_vpred(&mut uws, 8, 1, 1, stride);
+                predict_vpred(&mut vws, 8, 1, 1, stride);
+            },
+            ChromaMode::H => {
+                predict_hpred(&mut uws, 8, 1, 1, stride);
+                predict_hpred(&mut vws, 8, 1, 1, stride);
+            },
+            ChromaMode::TM => {
+                predict_tmpred(&mut uws, 8, 1, 1, stride);
+                predict_tmpred(&mut vws, 8, 1, 1, stride);
+            },
+        }
+
+        for y in 0usize..2 {
+            for x in 0usize..2 {
+                let i = x + y * 2;
+                let mut urb = [0i32; 16];
+                urb.copy_from_slice(&resdata[16 * 16 + i * 16..16 * 16 + i * 16 + 16]);
+
+                let y0 = 1 + y * 4;
+                let x0 = 1 + x * 4;
+                add_residue(&mut uws, &urb, y0, x0, stride);
+
+                let mut vrb = [0i32; 16];
+                vrb.copy_from_slice(&resdata[20 * 16 + i * 16..20 * 16 + i * 16 + 16]);
+
+                add_residue(&mut vws, &vrb, y0, x0, stride);
```

Comment:
```suggestion
                let urb: &[i32; 16] = resdata[16 * 16 + i * 16..][..16].try_into().unwrap();

                let y0 = 1 + y * 4;
                let x0 = 1 + x * 4;
                add_residue(&mut uws, urb, y0, x0, stride);

                let vrb: &[i32; 16] = resdata[20 * 16 + i * 16..][..16].try_into().unwrap();

                add_residue(&mut vws, vrb, y0, x0, stride);
```

This removes two mem copies. I know the `.try_into()` looks bad but it can never fail and the compiler optimizes it away.

You'll also have to add `use std::convert::TryInto;` at the beginning of the file. I can't seem to leave a code suggestion for it.

Reply:
Will also do this for the luma one then too.